In [2]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

In [43]:
#load data
mnist = input_data.read_data_sets("MNIST_data", one_hot=True)

batch_size = 100

batch_count = mnist.train.num_examples // batch_size

#define place holder
x = tf.placeholder(tf.float32, [None, 784])
y = tf.placeholder(tf.float32, [None, 10])

#create a 1-level neural network
W = tf.Variable(tf.zeros([784, 10]))
b = tf.Variable(tf.zeros([10]))
prediction = tf.nn.softmax(tf.matmul(x, W)+b)

#二次代价函数
# loss = tf.reduce_mean(tf.square(y - prediction))
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y, logits=prediction))
# loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=y, logits=prediction))

#train
train = tf.train.GradientDescentOptimizer(0.2).minimize(loss)

init = tf.global_variables_initializer()

#比较预测值与实际值
correct_pred = tf.equal(tf.argmax(y, 1), tf.argmax(prediction, 1)) #argmax 返回沿着某个维度最大值的位置

#准确率
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

with tf.Session() as sess:
    sess.run(init)
    
    for step in range(20):
        for batch in range(batch_count):
            batch_xs, batch_ys = mnist.train.next_batch(batch_size)
            sess.run(train, feed_dict={x:batch_xs, y:batch_ys})
        
        acc = sess.run(accuracy, feed_dict={x:mnist.test.images, y:mnist.test.labels})
        print("step:" + str(step) + " accurancy:" + str(acc))




Extracting MNIST_data\train-images-idx3-ubyte.gz
Extracting MNIST_data\train-labels-idx1-ubyte.gz
Extracting MNIST_data\t10k-images-idx3-ubyte.gz
Extracting MNIST_data\t10k-labels-idx1-ubyte.gz
step:0 accurancy:0.8239
step:1 accurancy:0.8901
step:2 accurancy:0.9007
step:3 accurancy:0.9057
step:4 accurancy:0.9087
step:5 accurancy:0.9104
step:6 accurancy:0.9122
step:7 accurancy:0.9132
step:8 accurancy:0.9152
step:9 accurancy:0.9159
step:10 accurancy:0.9171
step:11 accurancy:0.918
step:12 accurancy:0.9188
step:13 accurancy:0.9198
step:14 accurancy:0.9203
step:15 accurancy:0.9201
step:16 accurancy:0.9199
step:17 accurancy:0.9204
step:18 accurancy:0.9216
step:19 accurancy:0.9211





cross entropy:
step:0 accurancy:0.8239
step:1 accurancy:0.8901
step:2 accurancy:0.9007
step:3 accurancy:0.9057
step:4 accurancy:0.9087
step:5 accurancy:0.9104
step:6 accurancy:0.9122
step:7 accurancy:0.9132
step:8 accurancy:0.9152
step:9 accurancy:0.9159
step:10 accurancy:0.9171
step:11 accurancy:0.918
step:12 accurancy:0.9188
step:13 accurancy:0.9198
step:14 accurancy:0.9203
step:15 accurancy:0.9201
step:16 accurancy:0.9199
step:17 accurancy:0.9204
step:18 accurancy:0.9216
step:19 accurancy:0.9211